compiler/z/codegen/OMRMemoryReference.cpp

```diff
@@ -578,16 +578,26 @@ OMR::Z::MemoryReference::MemoryReference(TR::Node * rootLoadOrStore, TR::CodeGen
          self()->createUnresolvedDataSnippet(rootLoadOrStore, cg, symRef, tempReg, isStore);
          }
 
+
+      // VSI vector load/store instructions, such as VSTRL, VLRL, and VSTEB, lack the I field and do not support
+      // index registers. As a result of this, we need to set separate index register so that an intermediate LA instruction
+      // is emitted to add up the index and base registers.
+      bool isPackedDecimalVectorLoadStore = rootLoadOrStore->getType().isAnyPacked()
+              && (TR::Compiler->target.cpu.getS390SupportsVectorPackedDecimalFacility() && !TR::Options::getCmdLineOptions()->getOption(TR_DisableVectorBCD))
+              && (rootLoadOrStore->getOpCode().isLoad() || rootLoadOrStore->getOpCode().isStore());
+
       // indexRegister may setup in populateMemoryReference to hold the address for the load/store
       // since indexRegister is invalid in a RS instruction (LM/STM for long), we need to insert an intermediate
       // LA instruction to get the address using index and base register before doing LM/STM
       // we'll never generate LM to load/store long int on 64bit platform
       // storageReference cases may contain symbol sizes from 1->64 and in any case this condition is now enforced when
       // the subsequent generate*Instruction routine is called with this memRef
-      if ((rootLoadOrStore->getType().isInt64() || (symbol && symbol->getSize() == 8)) &&
-         !(TR::Compiler->target.is64Bit() || cg->use64BitRegsOn32Bit()) &&
-          !canUseRX &&
-          (storageReference == NULL))
+      bool separateIndexRegisterForRSLoadStore = (rootLoadOrStore->getType().isInt64() || (symbol && symbol->getSize() == 8)) &&
+                                                 !(TR::Compiler->target.is64Bit() || cg->use64BitRegsOn32Bit()) &&
+                                                 !canUseRX &&
+                                                 (storageReference == NULL);
+
+      if (isPackedDecimalVectorLoadStore || separateIndexRegisterForRSLoadStore)
```

Comment:
I don't think this will work. For `LM` this works because there exists a long displacement `LMY` alternative. From `VSTRL` for example there does not exist a long displacement version. What happens if the displacement is more than 4096? 

I'd like to see a test case for this via Tril. I suspect this implementation will fail in undefined ways.

In my opinion this type of code does not belong in the MemoryReference constructor. The constructor should in theory not care about instruction formats. It should create the best possible memory reference given a node. If the instruction in which the memory reference will be used has certain restrictions, such as `VSTRL`, the `generateVSIInstruction` API should handle such restrictions.

See `generateRSInstruction` for example which supports the exact same memory reference format as `VSI` instructions:

https://github.com/eclipse/omr/blob/03270bf3b9e51e02c22fc8e6c5b88c60d928db7c/compiler/z/codegen/S390GenerateInstructions.cpp#L1086-L1100

We need to also be calling `enforce4KDisplacementLimit` to fixup the displacement if needed. See `enforceSSFormatLimits` as a relevant example.